wallet/wallet.c

```diff
@@ -1640,3 +1640,28 @@ bool wallet_network_check(struct wallet *w,
 	}
 	return true;
 }
+
+void wallet_block_add(struct wallet *w, struct block *b)
+{
+	sqlite3_stmt *stmt = db_prepare(w->db,
+					"INSERT INTO blocks "
+					"(height, hash, prev_hash) "
+					"VALUES (?, ?, ?);");
+	sqlite3_bind_int(stmt, 1, b->height);
+	sqlite3_bind_sha256_double(stmt, 2, &b->blkid.shad);
+	if (b->prev) {
+		sqlite3_bind_sha256_double(stmt, 3, &b->prev->blkid.shad);
+	}else {
+		sqlite3_bind_null(stmt, 3);
+	}
+	db_exec_prepared(w->db, stmt);
+}
+
+void wallet_block_remove(struct wallet *w, struct block *b)
+{
+	sqlite3_stmt *stmt = db_prepare(w->db, "DELETE FROM blocks "
+					"WHERE hash = ? OR height >= ?");
```

Comment:
The `OR` in the condition means the equality check on `hash` does not buy you any additional assurance of correctness.  It will only be relevant if the `hash` and `height` of the given `struct block *` does not match the entry in the DB, but if your `struct block *` does not match the corresponding entry in the DB you have bigger problems (a cache incoherence).  If the `hash` and `height` on the `struct block *` maches the DB row, then `height >= ?` is sufficient.  But that then means that `wallet_block_rollback` is sufficient and `wallet_block_remove` is unnecessary.

Reply:
Right, I should remove the height check, I was hoping to replicate the `remove_tip` logic, hoping that it'd remove in reverse height order.